pkg/limayaml/validate.go

```diff
@@ -177,7 +177,14 @@ func validateNetwork(yNetwork Network) error {
 		// The field is called VDE.VNL in anticipation of QEMU upgrading VDE2 to VDEplug4,
 		// but right now the only valid value on macOS is a path to the vde_switch socket directory,
 		// optionally with vde:// prefix.
-		if !strings.Contains(vde.VNL, "://") || strings.HasPrefix(vde.VNL, "vde://") {
+		// TODO: use networks.LimaSchema after solving circular dependency
+		if strings.HasPrefix(vde.VNL, "lima://") {
```

Comment:
Sorry, I still think that we shouldn't contaminate the VNL namespace.
`lima://` is reserved for a potential VDE4 plugin that would be called "lima".

https://github.com/lima-vm/lima/issues/175#issuecomment-907589748

Reply:
Ok, how do you want to implement it. As a separate list under `network`, e.g.

```yaml
network:
  lima:
  - shared
  - host
```

Reply:
SGTM

Reply:
It will have to be:

```yaml
network:
  lima:
  - network: bridged
    macAddress: 11:22:33:44:55:66
    name: lima0
  - network: host
```

etc. It is unfortunate that we use `vde.name` for the interface name. I wish it was `interface`, so we could use `lima.name` and `lima.interface` instead of `lima.network` and `lima.name`, but I guess it is too late to change this now.

Any strong feelings about the field names?

Reply:
I don't have a strong opinion.

If we want to change `name` to `interface`, we can safely do that by deprecating `name` with a warning message.

Reply:
This turns out to be really painful to implement, as we would have to duplicate so much code in validation. And the code would still not be independent as we would still have to verify that there are no duplicate interface names between the `VDE` and the `lima` lists.

So if we can still change the `network` layout, then I would suggest going with:

```yaml
network:
- vnl: /var/run/vde.ctl
  interface: vde0
- lima: shared
  interface: lima0
```

Either `vnl` or `lima` would be required, and `switchPort` would only be valid when you use `vnl`.

The problem is that we can't really do a proper deprecation message, as `network` would turn from a `map` to a `slice`.

I still think reserving the `lima://` scheme for our internal use is the most elegant solution, as I think it will be rather unlikely that there ever will be a VDE4 plugin called `lima`. But if we can change the `network` layout, then I'm fine with that too.

Reply:
Ok, one more thought: we could call the new structure `networks`, and deprecate `network`.

Reply:
> we could call the new structure `networks`, and deprecate `network`.

I've implemented this in https://github.com/lima-vm/lima/pull/239/commits/03e90a2 (separate commit for review; will squash later).

It throws an error when you use both the new and the deprecated fields:

```
$ l start alpine.yaml
? Creating an instance "alpine" Proceed with the default configuration
FATA[0001] the YAML is invalid, saved the buffer as "lima.REJECTED.yaml": you cannot use deprecated field `network.VDE` together with replacement field `networks`
```

And will show a deprecation warning when the file uses the old field:

```
$ l start alpine.yaml
? Creating an instance "alpine" Proceed with the default configuration
WARN[0001] field `network.VDE` is deprecated; please use `networks` instead
WARN[0001] field `network.VDE` is deprecated; please use `networks` instead
WARN[0001] field `network.VDE` is deprecated; please use `networks` instead
WARN[0001] field `network.VDE` is deprecated; please use `networks` instead
WARN[0001] field `network.VDE` is deprecated; please use `networks` instead
WARN[0001] public key "/Users/jan/.ssh/foo.pub" doesn't seem to be in ssh format
INFO[0001] Attempting to download the image from "https://github.com/lima-vm/alpine-lima/releases/download/v0.1.4/alpine-lima-std-3.13.5-x86_64.iso"
...
```

The warning is show multiple times because the file is validated several times. I'm somewhat surprised it happens 5 times, but I think it is unavoidable to do it multiple times, unless we cache all the instance data returned by `store.Inspect`.

Haven't done the other changes (network validation, metrics, help), but let me know if you think this is now an acceptable way to declare the networks for an instance!

This is now a working minimal network config:

```yaml
networks:
- lima: shared
```

And I changed the default interface names from `vde0` etc. to `lima0` etc because the use of vde is kind of an implementation details now.